### ISSUE TITLE
Bind the viewer in the napari.gui_qt() context manager

### ISSUE THREAD

**Issue by HagaiHargil** [napari/napari#900]
## ❓ Questions and Help

I was wondering why can't I do the following:

```python
with napari.gui_qt() as viewer:
   viewer.add_image(...)
```

It seems l like a nice QOL improvement, and a good default behavior for the `gui_qt` context manager. Also, I find that one of the most confusing things for new users is the duplication between `add_image` and `view_image` - they're never sure when to use which and why. I believe that if the context manager can return a viewer instance, all of the `view_X` methods will become mostly redundant since most use cases involving napari use this context manager for starting the Qt app.

**Comment by jni** [comment]
> most use cases involving napari use this context manager for starting the Qt app.

This is actually not true. I use napari mostly in IPython with `%gui qt`, and in that case, it's much faster to use `napari.view_image(my_array)` to look at it quickly, than to have to first create a `Viewer` object.

Regarding creating the Viewer and yielding it... Well, then I think the context manager should change its name. But, we did have discussions about this at one point, I can't remember why we steered away from yielding a fully created viewer. @sofroniewn @kne42 ?

**Comment by tlambert03** [comment]
I agree: `gui_qt()` does exactly what it looks like it should do: it creates a qt event loop for the gui (really _any_ gui... not just napari), and if we were going to yield a viewer it should be a different context manager with a new name (or a perhaps a `with_viewer=True` argument to `gui_qt`)

**Comment by sofroniewn** [comment]
> Also, I find that one of the most confusing things for new users is the duplication between add_image and view_image - they're never sure when to use which and why. I believe that if the context manager can return a viewer instance, all of the view_X methods will become mostly redundant since most use cases involving napari use this context manager for starting the Qt app.

I agree with this, I still don't love it - but as noted above for people that want a one liner we need it. I'm def open to more suggestions though, and having some input from new comers can be really helpful in these discussions.

> if we were going to yield a viewer it should be a different context manager with a new name (or a perhaps a with_viewer=True argument to gui_qt)

I'm open to this, but am wary of providing too many different options to all do the same thing. We'd also have to make sure it was clear then how to manipulate all the viewer properties that are now keyword arguments to the `Viewer`

**Comment by tlambert03** [comment]
> I'm open to this, but am wary of providing too many different options to all do the same thing. We'd also have to make sure it was clear then how to manipulate all the viewer properties that are now keyword arguments to the `Viewer`

yeah, I don't love it either, and wouldn't personally vote for that change.  I am receptive to the confusion that @HagaiHargil is reporting though, just not immediately sure what a next step would look like, since the confusion seems to come down to understanding the difference between a viewer method (`add_image`) and a global convenience function (`view_image`).

**Comment by HagaiHargil** [comment]
I agree that the current name is Qt-oriented, but I also think that 99% of users in 99% of use cases don't care about the Qt event loop - they just want a viewer handed to them. Moreover, most context managers they've dealt with in the past had the `as var:` clause, so they probably expect to be able to bind the viewer to a variable in the `with` line.

I think my proposed context manager should be a new function\class in the main napari namespace, and having the docs reference that function in the "getting started" part. The older `gui_qt` function can still be found in the more advanced parts of the tutorial for power users who wish to modify the viewer before its creation, or who wish to customize napari to their own specific use cases.

**Comment by sofroniewn** [comment]
> I agree that the current name is Qt-oriented, but I also think that 99% of users in 99% of use cases don't care about the Qt event loop - they just want a viewer handed to them. Moreover, most context managers they've dealt with in the past had the as var: clause, so they probably expect to be able to bind the viewer to a variable in the with line.

This really resonates with me. I'd be open to leaving the `gui_qt` for those that want it (can always slowly deprecate) but adding something like:
```python
with napari.viewer_context(ndisplay=3) as viewer:
   viewer.add_image(...)
```
where the viewer keyword args get passed to the `viewer_context` and introduce that first in the `getting started` section of our docs. Suggestions to an improved name to `viewer_context` are welcome. Also curious if people like being able to pass those keyword args to it or not ...

If would be nice if we autopopluated the console with the variable they use there and not always default to `viewer` too

**Comment by HagaiHargil** [comment]
This looks great, I like the keyword arguments idea.

> If would be nice if we autopopluated the console with the variable they use there and not always default to viewer too

I was actually about to create an issue for that today, but decided against it since I already opened three and pushed a new PR :X

**Comment by jni** [comment]
Yeah, the new context manager should definitely allow customising the viewer object that is yielded!

@HagaiHargil feel free to open as many issues as needed! It's very useful to keep track of things, even when they seem overwhelming. =) Having said that, I thought there was an issue about this already? Though maybe it was just an image.sc post...

**Comment by HagaiHargil** [comment]
Here's an idea for the context manager: How about we make the `Viewer` class a context manager in itself? Then you "re-export" the viewer in the main napari namespace to achieve:

```python
with napari.Viewer(ndisplay=3, ...) as viewer:
    viewer.add_image(img)
```

I believe this solution shouldn't break anything and is quite clear and simple from the user's POV - you get a viewer by instantiating a Viewer class. It also doesn't muddy the napari namespace with new functions tailored for a specific use-case \ issue I had.

**Comment by tlambert03** [comment]
yeah I like that syntax a lot.  so we would just be creating the qt app if it didn't exist in the `Viewer.__enter__` method?

There's a tiny part of me that still feels like we're just saying "many users don't know what qt is, so let's just put a different name on it (`Viewer`)..." but behind the scenes, the only thing the context manager is required for (the `__enter__` method) is exactly what `gui_qt` is already doing, just without calling it qt.  Yes, it's also instantiating a viewer... but that doesn't require a context manager.

I'm fine with it if everyone else likes it (I definitely think it's the right syntax) ... but it feels kiiiind of  like "same face, new name" ... 😊

**Comment by HagaiHargil** [comment]
> yeah I like that syntax a lot. so we would just be creating the qt app if it didn't exist in the Viewer.__enter__ method

That's the plan.

What I'm really trying to solve here is the duplication between `add_x` and `view_x`, which is a bigger pain point for new users (from the n=3 I've seen) than the `gui_qt()` context manager, which is both quite easy to use and kinda necessary due to qt.

**Comment by tlambert03** [comment]
yeah I see (and am sympathetic to) that confusion.  Do you propose that we actually remove the `napari.view_x` syntax altogether after adding this context manager?  Or leave it in, but remove it from all of our tutorials and examples?  Otherwise, we're just creating a third option, right?

**Comment by HagaiHargil** [comment]
I think it should eventually be removed. 

I'm not sure what is napari's deprecation policy, but if we emit a deprecation warning now it will give users (hopefully) enough time so that we could remove it by 0.3 or 0.4. We also need to remove it from tutorials, tests, etc...

**Comment by tlambert03** [comment]
still potentially ok with it, but it all just makes me a bit sad.  @HagaiHargil, I absolutely like both your motivations (decreasing confusion), and the proposed solution (a nice syntax) ... I'm just sad because it means we're removing what is a very nice and clean minimal functional syntax:

```python
from napari import view_image

view_image(array)
```

in favor of a context manager:
```python
from napari import Viewer

with Viewer() as viewer:
    viewer.add_image(array)
```

because of an understandable (but still extremely basic) misunderstanding about what it means to instantiate an object in python, and use one if its methods (e.g. if you already have a viewer, you can `add` an image **to** it, if you're starting from nothing, you `view` an image.) .  I _do_ want anyone at any level to be able to quickly use napari, _and_ for it to be as intuitive as possible... and if you've seen three people get confused by this, it's obviously an issue.  I'm on board with this proposal if every one else is, but it still feels a _tiny_ bit like a concession ...

_edit_:
and actually... along those lines, if we're going to remove `view_image` altogether to avoid confusion, unless we're actually doing some sort of setup and teardown (outside of the qt event loop which is already handled by `gui_qt`), why do we need this as a context manager instead of just:
```python
from napari import Viewer

viewer = Viewer()
viewer.add_image(array)
```
again, not opposed to any of it, just wanting to clarify my understanding of the goal here

**Comment by HagaiHargil** [comment]
I thought that the `gui_qt()` context manager is always needed in scripts, but not in IPython consoles. It seems fair to always use it as a context manager when writing code examples for new users, but obviously it's not a must.

My suggestion replaces `gui_qt()` with a context manager specifically designed to return a Viewer instance, you code snippet above will keep working wherever it now does.

Regarding your first concern - I agree that we lose a bit of brevity and conciseness by removing `view_*`.

**Comment by sofroniewn** [comment]
I think at the beginning we should add the

```python
with napari.Viewer(ndisplay=3, ...) as viewer:
``` 

and replace `gui_qt` in all the tutorials / docs (would we also leave `gui_qt` or get rid of it too? we probably would want to add a deprecation warning around that if we did)

I think removing the `view_*` entirely would be less desirable, until we think of a good one liner for the notebook use case.

Thoughts @jni?

**Comment by jni** [comment]
(a) I'm totally opposed to removing `view_image`. For interactive work, it's much nicer than creating a viewer and adding an image. We don't need the context manager at all in IPython with the Qt magic, so then it becomes extra clunky. Additionally, even though layers are one of the main selling points of napari, I *very often* just want to quickly look at an nD image, and keep going. In those cases, `view_image` is exactly what I want. I probably use `view_image` more than any other napari API.

The `view_*` for other layers, I use very rarely, though... We've had this discussion before. The argument for only having `view_image` was that napari is an image viewer first and foremost, so that's a special case. The argument against is that now you get even more inconsistency between `view_*` and `add_*`. I would be happy to revisit this one aspect. (Then we could go back to `napari.view()` which I liked. =) But, to be perfectly clear: I am *not* in favour of going down this route, I'm just not opposed to it. But I am totally opposed to removing *all* `view_` functions.

(b) as regards the viewer object acting as a context manager through `__enter__` and `__exit__`, if we can make it work identically, I'm quite happy to go for that. I had some issues before but I was probably doing it wrong. *Without* the `%gui qt` magic, you definitely need a context, @tlambert03. I *don't* think we should deprecate `gui_qt()` just yet — I want to see people using the new syntax in the wild for a bit before committing to that. I'm happy to "soft-deprecate", meaning remove most mentions of it in our docs. Once we're comfortable that `Viewer(...) as viewer` has people happy, we can deprecate it formally.

**Comment by tlambert03** [comment]
> Without the %gui qt magic, you definitely need a context, @tlambert03 

Did I say something that made you think I thought otherwise?  I dont mind using `Viewer.__enter__` to replace `gui_qt`;  but I hope we don’t start seeing people use the context manager everywhere (instead of just instantiating a viewer) even when a qt event loop has already been started, just because they no longer realize that was the only additional thing the context is doing for them. (Even if it’s technically harmless)

**Comment by jni** [comment]
@tlambert03 

Ah ok, I get now what you were going for with:

> unless we're actually doing some sort of setup and teardown (outside of the qt event loop which is already handled by gui_qt), why do we need this as a context manager instead of just:

Indeed, now the context manager logic might be *more* confusing rather than less: why sometimes create a viewer in a context and other times just making it? :thinking: And btw I'm not *entirely* sure that it's harmless to use the context when a Qt loop already exists...

**Comment by tlambert03** [comment]
> why sometimes create a viewer in a context and other times just making it?

Yes, this exactly.

I’m tempted to say that if we’re going to have a method or function that both creates a qt loop and instantiates a viewer, it should require both terms, such as `gui_qt(with_viewer=True)` or `Viewer(..., start_qt=True)`  Where the default on each is false

**Comment by HagaiHargil** [comment]
Just FYI, using the context manager inside IPython works flawlessly from what I could tell. Also, I'm less concerned about using the API "correctly" while having the precise mental model of how napari is integrated in the qt event loop, and more about giving the user the best possible usage experience. If this means having complex logic in `Viewer.__enter__` to make it work, while adding a bit of "magic" to napari's instantiation methods - I think it's completely acceptable.

**Comment by tlambert03** [comment]
yep, understood. thanks @HagaiHargil.  we know the context manager works in IPython, and we understand that the core proposal here is ultimately aimed at improving user experience.  It's just definitely worth talking about how best to do it.

**Comment by jni** [comment]
Of all the methods discussed `with Viewer() as v:` is my favourite. I don't love `gui_qt(with_viewer=True)` because now you have to have `viewer_kwargs` as an option, which is clunky. The potential for confusion about context vs no-context is real, but I think it can be resolved with good docs and error messages.

**Comment by sofroniewn** [comment]
I also like `with Viewer() as v:` best. Would I still be able to do `viewer = Viewer()` inside a notebook and have it work like normal too?

**Comment by tlambert03** [comment]
yep, adding the `__enter__` method wouldn't prevent normal instantiation.

**Comment by sofroniewn** [comment]
Ok I think we might be nearing consensus that for the next release we will add the with Viewer syntax to the code base and change our scripting examples in the documentation to use it, but we won’t deprecate or remove any other APIs at this point.

How does that sound to everyone?

**Comment by kne42** [comment]
I like the idea of this, but I am not sure that it will be so trivial.

The `gui_qt` context manager is equivalent to:
```python
app = QApplication(sys.argv)
viewer = Viewer(...)
...
app.exec_()
```

However, when using `Viewer` as a context manager, it becomes equivalent to this instead:
```python
viewer = Viewer(...)
app = QApplication(sys.argv)
...
app.exec_()
```

Because all of the Qt stuff is inside `Viewer.__init__`, it will be called before the app is instantiated and cause an error:
https://github.com/napari/napari/blob/7a04984472db9a0b74c4efc26d56a6f56b4ce9dd/napari/viewer.py#L35-L49

We could try instantiating the app if it does not exist already, but this could lead to confusion when people try to create a viewer without starting the event loop, since no error message would be shown (although we can combat this by adding a 3rd way to start the event loop with e.g. `viewer.show()`).

**Comment by HagaiHargil** [comment]
@tlambert03 offered here https://github.com/napari/napari/issues/900#issuecomment-580301487 the interesting idea of ditching the context manager all together. Do we really need it? Can we tear down the Qt event loop without it? Do we even need to worry about a "stray" event loop? I have very little knowledge in this area.

**Comment by tlambert03** [comment]
sorry for the confusion... I wasn't suggesting we could ditch context managers all together there.  In the first code block (`view_image(array)`), I was assuming that a qt event loop had already been started using either `ipython --gui=qt`, or a previous `%gui qt` magic.  The last code block also made that same assumption.

The reason we "need" the context manager is just in the scenario when you have started the console _without_ a gui event loop (i.e. calling `ipython` _without_ `--gui=qt`) and you don't want to use `%gui qt` for whatever reason.  Short of those two things, we have to have something like the `gui_qt()` context manager (or, require every user to make their own `QApplication()` instance and `exec_` it... which nobody wants to have to do).

**Comment by tlambert03** [comment]
in case you're curious to learn more about the event loop, there's a good intro [here](https://wiki.qt.io/Threads_Events_QObjects#Events_and_the_event_loop).  The point is, there needs to be some object running in an infinite loop in the background that is continually asking "did something happen? .... did something happen?", such that when the user _does_ interact with something (e.g. clicks a button), the event loop can detect it and trigger the appropriate callback.

**Comment by kne42** [comment]
I believe that the context manager was introduced simply to offer a nicer syntactic sugar (see #134). As mentioned in my previous comment, the following API could also be introduced:

```python
viewer = Viewer(...)
...
viewer.show()  # this is blocking and should be called last
```

**Comment by HagaiHargil** [comment]
Thanks @tlambert03 for being patient with me, and @kne42 for you suggestions.

I've implemented a draft of this idea in #968.

**Comment by sofroniewn** [comment]
Closing as we are moving away from this syntax towards `napari.run` see #2056
